Mathlib/MeasureTheory/Measure/AddContent.lean

```diff
@@ -0,0 +1,184 @@
+    m (s ∪ t) = m s + m t := by
+  by_cases hs_empty : s = ∅
+  · simp only [hs_empty, Set.empty_union, addContent_empty, zero_add]
+  classical
+  have h := addContent_sUnion (m := m) (I := {s, t}) ?_ ?_ ?_
+  rotate_left
+  · simp only [coe_pair, Set.insert_subset_iff, hs, ht, Set.singleton_subset_iff, and_self_iff]
+  · simp only [coe_pair, Set.pairwiseDisjoint_insert, pairwiseDisjoint_singleton,
+      mem_singleton_iff, Ne.def, id.def, forall_eq, true_and_iff]
+    exact fun _ => h_dis
+  · simp only [coe_pair, sUnion_insert, sUnion_singleton]
+    exact hst
+  convert h
+  · simp only [coe_pair, sUnion_insert, sUnion_singleton]
+  · rw [sum_insert, sum_singleton]
+    simp only [Finset.mem_singleton]
+    refine fun hs_eq_t => hs_empty ?_
+    rw [← hs_eq_t] at h_dis
+    exact Disjoint.eq_bot_of_self h_dis
+
+section IsSetSemiring
+
+lemma addContent_eq_add_diffFinset₀_of_subset (hC : IsSetSemiring C)
+    (hs : s ∈ C) (hI : ↑I ⊆ C) (hI_ss : ∀ t ∈ I, t ⊆ s)
+    (h_dis : PairwiseDisjoint (I : Set (Set α)) id) :
+    m s = ∑ i in I, m i + ∑ i in hC.diffFinset₀ hs hI, m i := by
+  classical
+  conv_lhs => rw [← hC.sUnion_union_diffFinset₀_of_subset hs hI hI_ss]
+  rw [addContent_sUnion]
+  · rw [sum_union]
+    exact hC.disjoint_diffFinset₀ hs hI
+  · rw [coe_union]
+    exact Set.union_subset hI (hC.diffFinset₀_subset hs hI)
+  · rw [coe_union]
+    exact hC.pairwiseDisjoint_union_diffFinset₀ hs hI h_dis
+  · rwa [hC.sUnion_union_diffFinset₀_of_subset hs hI hI_ss]
+
+lemma sum_addContent_le_of_subset (hC : IsSetSemiring C)
+    (h_ss : ↑I ⊆ C) (h_dis : PairwiseDisjoint (I : Set (Set α)) id)
+    (ht : t ∈ C) (hJt : ∀ s ∈ I, s ⊆ t) :
+    ∑ u in I, m u ≤ m t := by
+  classical
+  rw [addContent_eq_add_diffFinset₀_of_subset hC ht h_ss hJt h_dis]
+  exact le_add_right le_rfl
+
+lemma addContent_mono (hC : IsSetSemiring C) (hs : s ∈ C) (ht : t ∈ C)
+    (hst : s ⊆ t) :
+    m s ≤ m t := by
+  have h := sum_addContent_le_of_subset (m := m) hC (I := {s}) ?_ ?_ ht ?_
+  · simpa only [sum_singleton] using h
+  · rwa [singleton_subset_set_iff]
+  · simp only [coe_singleton, pairwiseDisjoint_singleton]
+  · simp [hst]
+
+end IsSetSemiring
+
+section IsSetRing
+
+lemma addContent_union (hC : IsSetRing C) (hs : s ∈ C) (ht : t ∈ C)
+    (h_dis : Disjoint s t) :
+    m (s ∪ t) = m s + m t :=
+  addContent_union' hs ht (hC.union_mem hs ht) h_dis
+
+lemma addContent_union_le (hC : IsSetRing C) (hs : s ∈ C) (ht : t ∈ C) :
+    m (s ∪ t) ≤ m s + m t := by
+  rw [← union_diff_self, addContent_union hC hs (hC.diff_mem ht hs)]
+  · exact add_le_add le_rfl
+      (addContent_mono hC.isSetSemiring (hC.diff_mem ht hs) ht (diff_subset _ _))
+  · rw [Set.disjoint_iff_inter_eq_empty, inter_diff_self]
+
+lemma addContent_biUnion_le {ι : Type*} (hC : IsSetRing C) {s : ι → Set α}
+    {S : Finset ι} (hs : ∀ n ∈ S, s n ∈ C) :
+    m (⋃ i ∈ S, s i) ≤ ∑ i in S, m (s i) := by
+  classical
+  revert hs
+  refine Finset.induction ?_ ?_ S
```

Comment:
You can use `induction'` here to avoid reverting. Something like
```
  induction' S using Finset.induction with a S ha ih
```
Same thing elsewhere.